web/src/beta/features/PluginPlayground/Plugins/usePlugins.ts

```diff
@@ -170,42 +170,77 @@ export default () => {
     [selectedPlugin, selectedFileId]
   );
 
-  const handleFileUpload = useFileInput((fileList) => {
-    const file = fileList?.[0];
-    if (!file) {
-      return;
-    }
-    const result = validateFileTitle(
-      file.name,
-      selectedPlugin.files.map((f) => f.title)
-    );
-    if (!result.success) {
-      setNotification({ type: "error", text: result.message });
-      return;
-    }
+  const handleFileUpload = useFileInput(
+    (fileList) => {
+      const file = fileList?.[0];
+      if (!file) {
+        setNotification({ type: "error", text: "File not found" });
+        return;
+      }
 
-    const reader = new FileReader();
+      if (file.type !== "application/zip") {
+        setNotification({ type: "error", text: "Invalid file type" });
+        return;
+      }
 
-    reader.onload = async (event) => {
-      const body = event?.target?.result;
-      if (typeof body != "string") return;
-      const fileItem = {
-        id: uuidv4(),
-        title: file.name,
-        sourceCode: body
-      };
-      // Note: When a new file is uploaded, select that file
-      setSelectedFileId(fileItem.id);
-      setPlugins((plugins) =>
-        plugins.map((plugin) =>
-          plugin.id === selectedPlugin.id
-            ? { ...plugin, files: [...plugin.files, fileItem] }
-            : plugin
-        )
-      );
-    };
-    reader.readAsText(file);
-  });
+      const zip = new JSZip();
+      zip.loadAsync(file).then((zip) => {
+        const files = Object.values(zip.files).filter((file) => !file.dir);
+
+        if (files.length === 0) {
+          setNotification({ type: "error", text: "Zip file is empty" });
+          return;
+        }
+
+        // Prevent path traversal
+        if (files.some((file) => file.name.includes(".."))) {
+          setNotification({ type: "error", text: "Invalid file paths in zip" });
+          return;
+        }
+
+        const filePromises = files.map((file) => file.async("text"));
+
+        Promise.all(filePromises)
+          .then((fileContents) => {
+            const pluginFiles = fileContents.map((content, index) => ({
+              id: uuidv4(),
+              title: files[index].name.split("/")[1],
+              sourceCode: content
+            }));
+
+            const hasValidFiles = pluginFiles.every(
+              (file) =>
+                file.title.endsWith(".yml") || file.title.endsWith(".js")
+            );
+
+            if (!hasValidFiles) {
+              setNotification({
+                type: "error",
+                text: "Zip file must only contain .yml and .js files"
+              });
+              return;
+            }
+
+            const newPlugin = {
+              id: uuidv4(),
+              title: file.name,
+              files: pluginFiles
+            };
+
+            setPlugins((plugins) => [...plugins, newPlugin]);
+            setSelectedPluginId(newPlugin.id);
```

Comment:
This `setSelectedPlugin` can't work well since this new plugin is not exist on category, once you click `MyPlugin` it will revert to the previous MyPlugin.
Currently i'm think about making a new category `Imported`, as a dynamic category, similar to `Shared`, we can keep adding imported plugin into that category, and do select, what do you think?

Reply:
Okey!